### PR TITLE
Desugar common foreach loop to while loops

### LINKED ISSUE #41952: [Improvement]: Reduce the overhead of using `foreach` loop
### Description

Consider the fallowing example,
```
function feLoop(int[] values) returns int {
    int sum = 0;
    foreach int item in values {
        sum += item;
    }
    return sum;
}

function intLoop(int[] values) returns int {
    int sum = 0;
    int len = values.length();
    foreach int i in  0..<len {
        sum += values[i];
    }
    return sum;
}

function whileLoop(int[] values) returns int {
    int sum = 0;
    int len = values.length();
    int i = 0;
    while (i < len) {
        sum += values[i];
        i += 1;
    }
    return sum;
}
```
Execution time for each of these functions are (length of values in 100000),
+ feLoop - 0.207989991 ns
+ intLoop -  0.103210340 ns
+ whileLoop - 0.008965463 ns

This means using `foreach` loop here is significantly slower than `while`. Look into why this is the case (most likely this has to do with how we implement our iterators) and try to reduce the gap between `while` and `foreach`

### Describe your problem(s)

_No response_

### Describe your solution(s)

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Reduce the overhead of using `foreach` loops for common cases
Fixes #41952

## Approach
Desugar `foreach` loops over range expression and lists in to while loops without using iterators

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 1]
```diff
@@ -5019,6 +5021,86 @@ public void visit(BLangForeach foreach) {
         }
     }
 
+    private boolean canEliminateIterator(BLangForeach loop) {
+        BLangExpression collection = loop.collection;
+        return switch (collection.getKind()) {
+            case BINARY_EXPR -> true;
+            case SIMPLE_VARIABLE_REF -> collection.expectedType.getKind() == TypeKind.ARRAY;
```
Comment (LakshanWeerasinghe) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:5028]: why can't we extend this to support list-constructors and other expressions where the expected type is an array?

Comment (heshanpadmasiri) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 2]
```diff
@@ -5019,6 +5021,88 @@ public void visit(BLangForeach foreach) {
         }
     }
 
+    private boolean canEliminateIterator(BLangForeach loop) {
+        BLangExpression collection = loop.collection;
+        return switch (collection.getKind()) {
+            case BINARY_EXPR, LIST_CONSTRUCTOR_EXPR -> true;
+            case SIMPLE_VARIABLE_REF -> {
```
Comment (LakshanWeerasinghe) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:5028]: We can also include function call expressions, let expressions and query expressions.

File: tests/jballerina-unit-test/src/test/resources/test-src/jvm/foreach-arrays.bal [hunk 3]
```diff
@@ -239,6 +239,66 @@ function testArrayWithNullElements() returns string {
...
+    string[] values = ["a", "b", "c"];
+    int sum = 0;
+    foreach string entity in values {
+
+        string[] relationFields = from string each in values
+            where entity + each != "foo"
+            select each;
+        sum += relationFields.length();
+    }
+    assertEquality(sum, 9);
```
Comment (nipunayf) [inline tests/jballerina-unit-test/src/test/resources/test-src/jvm/foreach-arrays.bal:252]:
```suggestion
    assertEquality(9, sum);
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/jvm/ForeachArrayTests.java [hunk 4]
```diff
@@ -218,11 +219,41 @@ public void testArrayWithNullElements() {
...
     @Test
     public void testEmptyArray() {
         BRunUtil.invoke(program, "testEmptyArray");
     }
 
+    // These ensure observable behaviour of foreach statements are the same even when we optimize away the iterator
+    @Test
+    public void testMutatingArray() {
+        BRunUtil.invoke(program, "testMutatingArray");
+    }
```
Comment (nipunayf) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/jvm/ForeachArrayTests.java:246]: Can we move these to a DataProvider?

Comment (heshanpadmasiri) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/jvm/ForeachArrayTests.java]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 5]
```diff
@@ -5114,6 +5117,97 @@ public void visit(BLangForeach foreach) {
...
+        BLangExpression collection = loop.collection;
+        if (collection.getKind() == NodeKind.BINARY_EXPR) {
+            // range expression
+            return true;
+        }
+        TypeKind kind = collection.getBType().getKind();
+        return kind == TypeKind.ARRAY || kind == TypeKind.TUPLE;
+    }
+
+    private BLangBlockStmt desugarForeachToWhileWithoutIterator(BLangForeach foreach) {
```
Comment (lochana-chathura) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:5130]: Is it possible to add a comment showing the before desugar and after desugar code similar to`desugarForeachToWhileWithIterator()` method?

https://github.com/ballerina-platform/ballerina-lang/blob/ceb5155986a87aa68cea016262534f9fe9aa097d/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java#L5315-L5340

Comment (heshanpadmasiri) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:5156]: Fixed

**Review by lochana-chathura — APPROVED** [review]
LGTM

**Pull request merged**